composer/index.ts

```diff
@@ -0,0 +1,168 @@
+import sqlite3 from "sqlite3";
+import { open } from "sqlite";
+import { promisify } from "util";
+import toml from "@iarna/toml";
+import yargs from "yargs";
+
+const execFile = promisify(require("node:child_process").execFile);
+const writeFile = promisify(require("node:fs").writeFile);
+const unlink = promisify(require("node:fs").unlink);
+
+const DB = "combinations.db";
+// Set the builder as docker generic for every group.
+const BUILDER = "docker:generic";
+
+// Command line arguments.
+const argv = yargs(process.argv.slice(2))
+    .options({
+        "git-ref": { type: "string", demandOption: true },
+        "git-target": { type: "string", demandOption: true },
+        total_instances: { type: "number", demandOption: true },
+    })
+    .parseSync();
+
+// TOML schema to generate.
+class Instance {
+    constructor(public count: number) {}
+}
+
+class BuildArgs {
+    TRANSPORT: string;
+    MUXER: string;
+    VERSION: string;
+    GIT_REF?: string;
+    GIT_TARGET?: string;
+
+    constructor(
+        transport: string,
+        muxer: string,
+        version: string,
+        gitRef: string,
+        gitTarget: string
+    ) {
+        this.TRANSPORT = transport;
+        this.MUXER = muxer;
+        this.VERSION = version;
+
+        if (version.includes("master")) {
+            this.GIT_REF = gitRef;
+            this.GIT_TARGET = gitTarget;
+        }
+    }
+}
+
+class BuildConfig {
+    constructor(build_args: BuildArgs) {}
+}
+
+class Group {
+    constructor(
+        public id: string,
+        public builder: string,
+        public instances: Instance,
+        public build_config: BuildConfig
+    ) {}
+}
+
+class Run {
+    constructor(public id: string, public groups: Group[]) {}
+}
+
+class Global {
+    constructor(
+        public plan: string,
+        public plan_case: string,
+        public total_instances: number
+    ) {}
+}
+
+class Composition {
+    constructor(public global: Global, public runs: Run[]) {}
+}
+
+async function main() {
+    sqlite3.verbose();
+
+    // Call sqlite to process the csv resource files and generate a database.
+    // We call the sqlite process instead of doing it here cause
+    // the dot commands are interpreted by the sqlite cli tool not sqlite itself,
+    // and it is a lot faster parsing the csv's.
```

Comment:
that we can't call sqlite [dot commands](https://sqlite.org/cli.html#special_commands_to_sqlite3_dot_commands_) via the database driver, therefore we do it via `execFile` therefore instead of parsing the `csv` file with js, we do it with `sqlite` instead cause it's faster.

>Our CSVs are <10 lines long.

but they are expected to grow I assume.

Reply:
> > Our CSVs are <10 lines long.
> 
> but they are expected to grow I assume

I am guessing they would have to grow into the thousands for performance to become a problem and thus a driving factor for the design. I hope that we won't have CSV files that are 1000s of lines long!

> do you suggest importing, parsing and inserting in typescript land instead?

I think we should do whatever results in the better design. If we want to have a test for this, I think it will likely be easier if we don't need to set up dummy files for the tests but can pass data in as variables. That would suggest that we use SQL to insert the data which in turn means parsing the files in TS.